sema/arrays_dictionaries_test.go

```diff
@@ -1329,6 +1329,135 @@ func TestCheckResourceArrayMapInvalid(t *testing.T) {
 	assert.IsType(t, &sema.InvalidResourceArrayMemberError{}, errs[0])
 }
 
+func TestCheckArrayReduce(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		fun test() {
+			let x = [1, 2, 3]
+			let sum =
+				fun (acc: Int, x: Int): Int {
+					return acc + x
+				}
+
+			let y: Int = x.reduce(initial: 0, sum)
+		}
+
+		fun testFixedSize() {
+			let x : [Int; 5] = [1, 2, 3, 21, 30]
+			let product =
+				fun (acc: Int, x: Int): Int {
+					return acc * x
+				}
+
+			let y: Int = x.reduce(initial: 1, product)
+		}
```

Comment:
Add a test for calling the reduce function on an unauthorized reference to a struct-array (e.g: `&[S]`, where `S` is a struct type). And the `reduce` method takes a reducer-function having the signature as `fun (acc: Int, x: S): Int`, where it performs a mutation operation on the element `x`, such as `x.index = x.index +1` inside the reducer-function